gluster/fs/pkg/volume/delete.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volume
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+)
+
+func (p *glusterfsProvisioner) Delete(volume *v1.PersistentVolume) error {
+	var err error
+	class, err := GetClassForVolume(p.client, volume)
+	if err != nil {
+		glog.Errorf("Fail to get class for volume: %v", volume)
+		return err
+	}
+	cfg, err := NewProvisionerConfig(volume.Name, class.Parameters)
+	if err != nil {
+		return fmt.Errorf("Parameter is invalid: %s", err)
+	}
+
+	pvc := volume.Spec.ClaimRef
+	if pvc == nil {
+		glog.Errorf("glusterfs: ClaimRef is nil")
+		return fmt.Errorf("glusterfs: ClaimRef is nil")
+	}
+	if pvc.Namespace == "" {
+		glog.Errorf("glusterfs: namespace is nil")
+		return fmt.Errorf("glusterfs: namespace is nil")
+	}
+	p.deleteVolume(pvc.Namespace, pvc.Name, cfg)
+
+	//TODO ignorederror
+	err = p.allocator.Release(volume)
+	if err != nil {
+		glog.Errorf("glusterfs: error to release GID: %v", err)
+	}
+
+	return nil
+}
+
+func (p *glusterfsProvisioner) deleteVolume(
+	namespace string, name string, cfg *ProvisionerConfig) {
+
+	if cfg.DeletionPolicy == "delete" {
```

Comment:
IMO we can't do this, this goes against the meaning of the `Delete` reclaim policy on the PV. `Delete` must mean `delete PVC -> delete PV's storage asset -> delete PV (+endpoint+service)`. If the admin configures DeletionPolicy != "delete" then it becomes `delete PVC -> delete PV (+endpoint+service)` which may go against user's expectations, i.e. they expect their data to not be accessible by anyone anymore.

The alternative is to set `Retain` on the PV according to a storage class parameter, in which case Delete() will never be called on the PV and the PV will simply go to the `Released` state. https://github.com/kubernetes/kubernetes/pull/47987 will be coming in 1.8, so there could be potential conflicts between the paramater and storageClass.reclaimpolicy to resolve if we go this route.